### DISCUSSION TITLE
Adding more AsyncJob wrapper functions?

### DISCUSSION THREAD

**Discussion by dsisco11** [SteamRE/SteamKit#1095] [General]
Just wanted to ask about this before possibly wasting time.
Is there some reason that functions such as 'SteamUser.LogOn()' do not have an 'SteamUser.LogOnAsync()' variant which returns an AsyncJob?

**Comment by yaakov-h** [comment]
Jobs are a particular part of the protocol with certain semantics.

Unless we've missed something, functions that don't return AsyncJobs do not use Job messages and adapting the non-Job behaviour to TPL convenience is non-trivial at a library level, as we can't make certain assumptions that a consuming app could decide to make.

Logon messages in particular are not Job messages, and consumers need to be able to handle logon responses that do not have a corresponding logon request.

**Reply by dsisco11** [reply]
Ahh, I really should have known that they don't use the job system. I remember having that realization some years ago when going through the SteamKit source on a more thorough level.
I'll just structure around TaskCompletionSource's and Semaphores then.

However;
```consumers need to be able to handle logon responses that do not have a corresponding logon request.```
Could you elaborate a bit more on what you mean by this? What sorts of logon responses have no corresponding request?
Are you referring to messages that result from the Steam servers forcibly logging the account out, such as when one account instance boots another out?
If so, are there other such events you could remind me of? Its been a few years since I've touched this stuff.

**Reply by yaakov-h** [reply]
There's only one I can remember off the top of my head, which is if you connect to a CM and don't send a login message before the timeout (30s? 60s?) in which case you will get a non-protobuf ClientMsgLogOnResponse right before booting you off.

**Reply by Civ01** [reply]
How can I find out if particular function/request to Steam should be asynjob? For example, "GetServerList#1"
